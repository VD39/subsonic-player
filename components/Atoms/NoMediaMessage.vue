<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="['column', 'centerItems', $style.content]">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: var(--width-height-100);
}

.content {
  position: absolute;
  inset: 50% auto auto 50%;
  transform: translate(-50%, -50%);
}
</style>
