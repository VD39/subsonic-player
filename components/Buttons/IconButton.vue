<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: string;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: number;
    iconWeight?: IconWeight;
    is?: ButtonType;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 24,
    iconWeight: 'duotone',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      $style.iconButton,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        [$style.fullWidth]: fullWidth,
      },
    ]"
    :disabled="disabled"
    :type="type"
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="phIcon"
      :class="$style.icon"
      aria-hidden="true"
      :size="iconSize"
      :weight="iconWeight"
      :color="iconColor"
    />

    <span
      ref="text"
      :class="[
        $style.text,
        {
          'visually-hidden': !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.iconButton {
  @mixin align-center;

  position: relative;
  display: inline-flex;
  gap: var(--space-8);
  padding: var(--button-spacing);
  margin: 0;
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.fullWidth {
  width: 100%;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  @mixin ellipsis-overflow;

  display: block;
  margin-bottom: 1px;
}
</style>
