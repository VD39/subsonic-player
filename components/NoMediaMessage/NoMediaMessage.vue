<script setup lang="ts">
defineProps<{
  icon: string;
  message: string;
}>();
</script>

<template>
  <div :class="$style.noMediaMessage">
    <div :class="$style.content">
      <component :is="icon" :size="50" />

      <p>{{ message }}</p>
    </div>
  </div>
</template>

<style module>
.noMediaMessage {
  position: relative;
  height: 100%;
}

.content {
  @mixin align-center;

  position: absolute;
  inset: 50% auto auto 50%;
  flex-direction: column;
  transform: translate(-50%, -50%);
}
</style>
