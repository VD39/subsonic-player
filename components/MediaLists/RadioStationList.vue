<script setup lang="ts">
import TrackImage from '@/components/TrackDetails/TrackImage.vue';
import DropdownMenu from '@/components/Dropdown/DropdownMenu.vue';
import DropdownItem from '@/components/Dropdown/DropdownItem.vue';
import DropdownDivider from '@/components/Dropdown/DropdownDivider.vue';
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';

defineProps<{
  radioStations: RadioStation[];
}>();

defineEmits([
  'addToQueue',
  'deleteRadioStation',
  'editRadioStation',
  'playStation',
]);
</script>

<template>
  <div v-if="radioStations.length" ref="radioStationWrapper">
    <div class="trackHeader">
      <div class="trackPlay">#</div>
      <div class="trackDetails">
        <p>Station</p>
      </div>
    </div>

    <ol>
      <li
        v-for="radioStation in radioStations"
        :key="radioStation.id"
        class="trackItem"
        data-test-id="radio-station"
      >
        <div class="trackPlay">
          <TrackImage
            :track-id="radioStation.id"
            @play-current-track="$emit('playStation', radioStation)"
          />
        </div>

        <div class="trackDetails">
          <a
            :href="radioStation.homePageUrl"
            target="_blank"
            :class="['clamp', $style.name]"
          >
            {{ radioStation.name }}
          </a>

          <DropdownMenu data-test-id="dropdown-menu">
            <DropdownItem
              data-test-id="edit-radio-station"
              @click="$emit('editRadioStation', radioStation)"
            >
              Edit station
            </DropdownItem>
            <DropdownItem
              data-test-id="delete-radio-station"
              @click="$emit('deleteRadioStation', radioStation)"
            >
              Delete station
            </DropdownItem>
            <DropdownDivider />
            <DropdownItem
              data-test-id="add-to-queue"
              @click="$emit('addToQueue', radioStation)"
            >
              Add to queue
            </DropdownItem>
            <DropdownItem
              data-test-id="play-station"
              @click="$emit('playStation', radioStation)"
            >
              Play station
            </DropdownItem>
          </DropdownMenu>
        </div>
      </li>
    </ol>
  </div>
  <NoMediaMessage v-else icon="PhRadio" message="No radio stations found." />
</template>

<style module>
.name {
  margin: 0 auto 0 0;
}
</style>
