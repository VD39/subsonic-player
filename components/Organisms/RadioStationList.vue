<script setup lang="ts">
import LinkOrText from '@/components/Atoms/LinkOrText.vue';
import MarqueeScroll from '@/components/Atoms/MarqueeScroll.vue';
import NoMediaMessage from '@/components/Atoms/NoMediaMessage.vue';
import DropdownDivider from '@/components/Molecules/Dropdown/DropdownDivider.vue';
import DropdownItem from '@/components/Molecules/Dropdown/DropdownItem.vue';
import DropdownMenu from '@/components/Molecules/Dropdown/DropdownMenu.vue';
import TrackPlayPause from '@/components/Organisms/TrackPlayPause.vue';

defineProps<{
  radioStations: RadioStation[];
}>();

defineEmits([
  'addToQueue',
  'deleteRadioStation',
  'editRadioStation',
  'playRadioStation',
]);
</script>

<template>
  <div v-if="radioStations.length" ref="radioStationWrapper" class="trackTable">
    <div class="trackHeader">
      <div class="trackCell">Stations</div>
      <div class="trackCell trackOptions" />
    </div>

    <div
      v-for="(radioStation, index) in radioStations"
      :key="radioStation.id"
      class="trackRow"
      data-test-id="radio-station"
    >
      <div class="trackCell">
        <div>
          <TrackPlayPause
            :track-id="radioStation.id"
            :image="radioStation.image"
            :track-number="index + 1"
            @play-track="$emit('playRadioStation', radioStation)"
          />

          <MarqueeScroll>
            <LinkOrText
              is="a"
              :is-link="!!radioStation.homePageUrl"
              :text="radioStation.name"
              :to="radioStation.homePageUrl"
              target="_blank"
            />
          </MarqueeScroll>
        </div>
      </div>

      <div class="trackCell trackOptions">
        <DropdownMenu ref="dropdownMenu">
          <DropdownItem
            is="a"
            v-if="radioStation.homePageUrl"
            :href="radioStation.homePageUrl"
            target="_blank"
          >
            Visit station
          </DropdownItem>
          <DropdownItem
            ref="editRadioStation"
            @click="$emit('editRadioStation', radioStation)"
          >
            Edit station
          </DropdownItem>
          <DropdownItem
            ref="deleteRadioStation"
            @click="$emit('deleteRadioStation', radioStation.id)"
          >
            Delete station
          </DropdownItem>
          <DropdownDivider />
          <DropdownItem
            ref="addToQueue"
            @click="$emit('addToQueue', radioStation)"
          >
            Add to queue
          </DropdownItem>
          <DropdownItem
            ref="playRadioStation"
            @click="$emit('playRadioStation', radioStation)"
          >
            Play station
          </DropdownItem>
        </DropdownMenu>
      </div>
    </div>
  </div>

  <NoMediaMessage
    v-else
    :icon="IMAGE_DEFAULT_BY_TYPE.radioStation"
    message="No radio stations found."
  />
</template>
