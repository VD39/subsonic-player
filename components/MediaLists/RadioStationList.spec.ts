import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { getFormattedRadioStationMock } from '@/test/helpers';
import IconButton from '@/components/Buttons/IconButton.vue';
import NoMediaMessage from '@/components/NoMediaMessage/NoMediaMessage.vue';
import RadioStationList from './RadioStationList.vue';

function factory(props = {}) {
  return mount(RadioStationList, {
    props: {
      radioStations: [],
      ...props,
    },
  });
}

describe('RadioStationList', () => {
  let wrapper: VueWrapper;

  beforeAll(() => {
    wrapper = factory();
  });

  it('matches the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe('when radioStations prop is an empty array', () => {
    it('does not show the radio station wrapper', () => {
      expect(wrapper.find({ ref: 'radioStationWrapper' }).exists()).toBe(false);
    });

    it('shows the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(true);
    });
  });

  describe('when radioStations prop is not an empty array', () => {
    const radioStations = getFormattedRadioStationMock(5);

    beforeAll(async () => {
      wrapper = factory({
        radioStations,
      });

      wrapper
        .findAllComponents('[data-test-id="dropdown-menu"]')[0]
        .findComponent(IconButton)
        .vm.$emit('click');

      await wrapper.vm.$nextTick();
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('shows the radio station wrapper', () => {
      expect(wrapper.find({ ref: 'radioStationWrapper' }).exists()).toBe(true);
    });

    it('shows the correct number of radio station items', () => {
      expect(wrapper.findAll('[data-test-id="radio-station"]').length).toBe(5);
    });

    it('does not show the NoMediaMessage component', () => {
      expect(wrapper.findComponent(NoMediaMessage).exists()).toBe(false);
    });

    describe('when the delete radio station button is clicked', () => {
      beforeAll(() => {
        (
          wrapper.findAllComponents(
            '[data-test-id="delete-radio-station"]',
          )[0] as VueWrapper
        ).vm.$emit('click');
      });

      it('emits the deleteRadioStation event with track', () => {
        expect(wrapper.emitted('deleteRadioStation')).toEqual([
          [radioStations[0]],
        ]);
      });
    });

    describe('when the edit radio station button is clicked', () => {
      beforeAll(() => {
        (
          wrapper.findAllComponents(
            '[data-test-id="edit-radio-station"]',
          )[0] as VueWrapper
        ).vm.$emit('click');
      });

      it('emits the editRadioStation event with track', () => {
        expect(wrapper.emitted('editRadioStation')).toEqual([
          [radioStations[0]],
        ]);
      });
    });

    describe('when the add to queue button is clicked', () => {
      beforeAll(() => {
        (
          wrapper.findAllComponents(
            '[data-test-id="add-to-queue"]',
          )[0] as VueWrapper
        ).vm.$emit('click');
      });

      it('emits the addToQueue event with track', () => {
        expect(wrapper.emitted('addToQueue')).toEqual([[radioStations[0]]]);
      });
    });

    describe('when the play station button is clicked', () => {
      beforeAll(() => {
        (
          wrapper.findAllComponents(
            '[data-test-id="play-station"]',
          )[0] as VueWrapper
        ).vm.$emit('click');
      });

      it('emits the playStation event with track', () => {
        expect(wrapper.emitted('playStation')).toEqual([[radioStations[0]]]);
      });
    });
  });
});
