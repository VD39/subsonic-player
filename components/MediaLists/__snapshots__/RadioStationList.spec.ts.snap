// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RadioStationList > matches the snapshot 1`] = `
"<div class="noMediaMessage">
  <div class="content">
    <phradio size="50"></phradio>
    <p>No radio stations found.</p>
  </div>
</div>"
`;

exports[`RadioStationList > when radioStations prop is not an empty array > matches the snapshot 1`] = `
"<div>
  <div class="trackHeader">
    <div class="trackPlay">#</div>
    <div class="trackDetails">
      <p>Station</p>
    </div>
  </div>
  <ol>
    <li class="trackItem" data-test-id="radio-station">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails"><a href="homepageUrl" target="_blank" class="clamp name">radio-station-0</a>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <div class="dropdownMenu">
              <ul class="menu">
                <li class="dropdownItem" data-test-id="edit-radio-station"><button class="iconButton fullWidth action action" type="button" data-test-id="edit-radio-station">
                    <!--v-if--><span class="text"> Edit station </span>
                  </button></li>
                <li class="dropdownItem" data-test-id="delete-radio-station"><button class="iconButton fullWidth action action" type="button" data-test-id="delete-radio-station">
                    <!--v-if--><span class="text"> Delete station </span>
                  </button></li>
                <li class="divider" aria-hidden="true"></li>
                <li class="dropdownItem" data-test-id="add-to-queue"><button class="iconButton fullWidth action action" type="button" data-test-id="add-to-queue">
                    <!--v-if--><span class="text"> Add to queue </span>
                  </button></li>
                <li class="dropdownItem" data-test-id="play-station"><button class="iconButton fullWidth action action" type="button" data-test-id="play-station">
                    <!--v-if--><span class="text"> Play station </span>
                  </button></li>
              </ul>
            </div>
          </transition-stub>
        </div>
      </div>
    </li>
    <li class="trackItem" data-test-id="radio-station">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails"><a href="homepageUrl" target="_blank" class="clamp name">radio-station-1</a>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
    </li>
    <li class="trackItem" data-test-id="radio-station">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails"><a href="homepageUrl" target="_blank" class="clamp name">radio-station-2</a>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
    </li>
    <li class="trackItem" data-test-id="radio-station">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails"><a href="homepageUrl" target="_blank" class="clamp name">radio-station-3</a>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
    </li>
    <li class="trackItem" data-test-id="radio-station">
      <div class="trackPlay">
        <div class="trackImage"><button class="iconButton itemLink playButton itemLink playButton" type="button" title="Play track">
            <phplay class="icon" aria-hidden="true" size="24" weight="duotone"></phplay><span class="text visually-hidden"> Play track </span>
          </button></div>
      </div>
      <div class="trackDetails"><a href="homepageUrl" target="_blank" class="clamp name">radio-station-4</a>
        <div class="dropdown" data-test-id="dropdown-menu"><button class="iconButton alignRight button button" type="button" title="More">
            <phdotsthreevertical class="icon" aria-hidden="true" size="24" weight="duotone"></phdotsthreevertical><span class="text visually-hidden">More</span>
          </button>
          <transition-stub name="fade" appear="false" persisted="false" css="true">
            <!--v-if-->
          </transition-stub>
        </div>
      </div>
    </li>
  </ol>
</div>"
`;
